<body>
  <div id="root"></div>
  <script src="https://unpkg.com/react@17/umd/react.development.js"></script>
  <script src="https://unpkg.com/react-dom@17/umd/react-dom.development.js"></script>
  <script src="https://unpkg.com/@babel/standalone/babel.min.js"></script>
  <script src="https://unpkg.com/prop-types@15.7.2/prop-types.js"></script>
  <style>
    img {
      max-width: 100%;
      border-radius: 8px;
    }
    button {
      display: flex;
      margin: 8px auto;
      background: yellowgreen;
      border: none;
      align-items: center;
      border-radius: 8px;
      padding: 4px 8px;
      cursor: pointer;
    }
    button.icon-refresh::before {
      content: '';
      display: inline-block;
      background-image: url('https://cdn0.iconfinder.com/data/icons/glyphpack/41/refresh-512.png');
      background-size: 25px;
      width: 25px;
      height: 25px;
      margin-right: 8px;
    }
  </style>

  <script type="text/babel">
    // https://www.bugsnag.com
    // https://sentry.io
    // https://www.npmjs.com/package/react-error-boundary

    const ErrorBoundaryContext = React.createContext();

    const useErrorHandling = () => {
      const contextValue = React.useContext(ErrorBoundaryContext);
      if (!contextValue) throw new Error('Provider!!!');
      return contextValue;
    };

    const ErrorSupplier = ({ children, FallBack }) => {
      const [hasError, setHasError] = React.useState(false);

      const value = React.useMemo(
        () => ({
          triggerError: () => setHasError(true),
          reset: setHasError.bind(null, false),
        }),
        [],
      );

      return (
        <ErrorBoundaryContext.Provider value={value}>
          {hasError ? <FallBack /> : children}
        </ErrorBoundaryContext.Provider>
      );
    };

    const RefreshButton = () => (
      <button className="icon-refresh" onClick={() => window.location.reload()}>
        refresh fallback
      </button>
    );

    class ErrorBoundary extends React.Component {
      // constructor(props) {
      //   super(props);
      //   this.state = { hasError: false };
      // }
      state = { hasError: false };
      // Update state so the next render will show the fallback UI.
      static getDerivedStateFromError(error) {
        console.log(`ðŸ‘‰ error`, error);
        console.dir(error.message);

        return { hasError: true, error };
      }

      render() {
        // if error
        if (this.state.hasError)
          // return <this.props.fallback error={this.state.error}/> ||(
          return (
            <div>
              <img src="https://images.kojaro.com/2018/4/c1f77b05-94e3-469e-8f37-fc23716b8e65.jpg" />
              <RefreshButton />
            </div>
          );

        return this.props.children;
      }
    }

    function ChildComponent() {
      const [counter, increment] = React.useReducer(s => s + 1, 0);
      React.useEffect(() => {
        // if (counter === 0) throw new Error('No, it can not be 4 either');
      }, [counter]);

      // 5 / 0

      if (counter === 5) throw new Error('counter must be <= 4');
      return (
        <div>
          <var
            onClick={() => {
              throw new Error('chert');
            }}
          >
            {counter}
          </var>
          <button onClick={increment} disabled={counter === 4}>
            increment
          </button>
        </div>
      );
    }

    function ComponentWithCatchError() {
      const errorHandling = useErrorHandling();
      const click = () => {
        try {
          throw new Error('chert');
        } catch (o_O) {
          console.log(`ðŸ‘‰ o_O`, o_O);
          errorHandling.triggerError();
        }
      };
      return (
        <div>
          <h1>ChildComponent</h1>
          <button onClick={click}>
            Click me to face error in chile component with catch
          </button>
          <img src="https://courses.w3schools.com/" onError={click} />
        </div>
      );
    }

    function Root() {
      return (
        <div className="App">
          <h1>Error Boundary</h1>
          <ErrorBoundary>
            <ChildComponent />
          </ErrorBoundary>
          <ErrorSupplier FallBack={RefreshButton}>
            <ComponentWithCatchError />
          </ErrorSupplier>
        </div>
      );
    }

    ReactDOM.render(<Root />, document.getElementById('root'));
  </script>
</body>
